# About
Need to convert a VCF into a gVCF so you can combine other data that wasn't generated by GATK? Well, 
this script is what you need.

Providing a VCF and a target BED file, this script will transform your plain single sample VCF into
a GVCF (sort of). Rather than doing it correctly (which would be a more complicated process), this approach
instead assumes that you have adequate coverage in your BED file. It will then add in your pre-specified 
values of GL, PL, GQ, and MIN_DP fields.


    usage: main.py [-h] -v VCF -b BED -f FASTA [-o OUTNAME] [-M MIN_DEPTH] [-P PL]
                   [-V {DEBUG,INFO,WARNING,ERROR,CRITICAL}]
    
    optional arguments:
      -h, --help            show this help message and exit
      -v VCF, --vcf VCF     VCF File to convert
      -b BED, --bed BED     Target BED file
      -f FASTA, --fasta FASTA
                            Where is the reference genome (FASTA) file?
      -o OUTNAME, --outname OUTNAME
                            What should the output file name be? [output.g.vcf]
      -M MIN_DEPTH, --min_depth MIN_DEPTH
                            Dummy value for min_depth [20]
      -P PL, --pl PL        Dummy value for PL (Non zero genotypes) [99]
      -V {DEBUG,INFO,WARNING,ERROR,CRITICAL}, --verbose {DEBUG,INFO,WARNING,ERROR,CRITICAL}
                            Set the logging level [INFO]

    

